frame/dutch-auction/src/lib.rs

```diff
@@ -1,313 +1,266 @@
-			for (order_id, order) in Orders::<T>::iter() {
-				match order.state {
-					AuctionState::AuctionStarted => {
-						if now > order.started + ONE_HOUR {
-							removed.push(order_id);
-						} else {
-							// for final protocol may be will need to transfer currency onto auction
-							// pallet sub account and send dex order with idempotency tracking id final protocol seems should include multistage lock/unlock https://github.com/paritytech/xcm-format or something
-							let delta_time = now - order.started;
-							let price: LiftedFixedBalance =
-								order.source_initial_price.initial_price.into();
-							let total_price = price.safe_mul(&order.source_total_amount.into())?;
-							let price = match order.function {
-								AuctionStepFunction::LinearDecrease(parameters) =>
-									parameters.price(total_price, delta_time),
-								AuctionStepFunction::StairstepExponentialDecrease(parameters) =>
-									parameters.price(total_price, delta_time),
-							}?
-							.checked_mul_int(1u64)
-							.ok_or(ArithmeticError::Overflow)?;
+		fn take(
+			from_to: &Self::AccountId,
+			order_id: Self::OrderId,
+			take: Take<Self::Balance>,
+		) -> Result<(), DispatchError> {
+			ensure!(take.is_valid(), Error::<T>::TakeParametersIsInvalid,);
+			let order = <SellOrders<T>>::try_get(order_id)
+				.map_err(|_x| Error::<T>::RequestedOrderDoesNotExists)?;
+			ensure!(
+				order.order.take.limit <= take.limit,
+				Error::<T>::TakeLimitDoesNotSatisfiesOrder,
+			);
+			let limit = order.order.take.limit.into();
+			// may consider storing calculation results within single block, so that finalize does
+			// not recalculates
+			let passed = T::UnixTime::now().as_secs() - order.added_at;
+			let _limit = order.configuration.price(limit, passed)?;
+			let quote = take.amount.safe_mul(&take.limit)?;
 
-							let price =
-								Price::<Self::GroupId, Self::Balance>::new_any(price.into());
+			T::MultiCurrency::reserve(order.order.pair.quote, from_to, quote)?;
+			<Takes<T>>::append(order_id, TakeOf::<T> { from_to: from_to.clone(), take });
 
-							let dex_order_intention = <T::Orderbook as Orderbook>::post(
-								&order.account_id,
-								order.source_asset_id,
-								order.target_asset_id,
-								order.source_total_amount,
-								price,
-								Permill::from_perthousand(5),
-							)?;
+			Ok(())
+		}
+	}
 
-							Orders::<T>::mutate(order_id, |order| {
-								// considers updating in place is
-								// safe during iteration
-								order.state = AuctionState::AuctionOnDex(dex_order_intention.id);
-							});
-						}
-					},
-					AuctionState::AuctionOnDex(_) => {
-						// waiting for off chain callback about order status
-						if now > order.started + ONE_HOUR {
-							removed.push(order_id);
-						}
-					},
-					_ => {
-						removed.push(order_id);
-					},
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(_n: T::BlockNumber) {
```

Comment:
Looks like can use `on_initialize` instead of `on_finalize` here.

Reply:
added comment